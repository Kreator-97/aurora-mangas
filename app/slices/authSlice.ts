import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { User } from '../../interfaces'

interface State {
  isLogged: Boolean;
  user    : User | null;
}

const initialState: State = {
  isLogged: false,
  user    : null,
}

export const authSlice = createSlice({
  name: 'auth',
  initialState,
  reducers: {
    login: (state, action:PayloadAction<{user: User}>) => {
      const { user } = action.payload
      return {...state, user, isLogged: true}
    },
  }
})

export const { login } = authSlice.actions
export default authSlice.reducer
