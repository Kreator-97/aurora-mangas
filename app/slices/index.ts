export { default as uiSlice } from './uiSlice'
export { default as authSlice } from './authSlice'
export { default as shoppingCartSlice } from './shoppingCartSlice'
