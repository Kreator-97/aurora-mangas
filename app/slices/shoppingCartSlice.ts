import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { ShoppingCart, ShoppingItem } from '../../interfaces'

const initialState: ShoppingCart = {
  items: [],
  total: 0
}

export const shoppingCartSlice = createSlice({
  name: 'shoppingCart',
  initialState,
  reducers: {
    addItem: (state, action: PayloadAction<ShoppingItem>) => {
      const { product, amount } = action.payload

      const item = state.items.find( item => item.product.id === product.id )

      if( !item ) {
        // if item is not included in previous items, we push it
        state.items.push({ product, amount })
        state.total = state.total + (product.price * amount )
        return
      }
      
      item.amount++
      state.total = state.total + (product.price * amount )
    },
    removeItem: (state, action: PayloadAction<{ id: string}>) => {
      const { id } = action.payload
      const items = state.items.filter( item => item.product.id !== id )
      
      state.items = items
      state.total = items.reduce( (acc, item) => (item.product.price * item.amount) + acc, 0)
    },
    setAmount: (state, action: PayloadAction<{amount:number, id: string}>) => {
      const { id, amount } = action.payload
      const index = state.items.findIndex( item => item.product.id === id )
      console.log(action.payload)
      if( index !== -1 ) {
        state.items[index].amount = amount
        state.total = state.items.reduce( (acc, item) => (item.product.price * item.amount) + acc, 0)
      }
    },
    setShoppingCart: (state, action: PayloadAction<ShoppingCart>) => {
      state = action.payload
      return state
    }
  }
})

export const { addItem, removeItem, setAmount, setShoppingCart } = shoppingCartSlice.actions
export default shoppingCartSlice.reducer
