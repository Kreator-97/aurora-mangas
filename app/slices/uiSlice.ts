import { createSlice } from '@reduxjs/toolkit'

interface State {
  isSidebarOpen: boolean;
  isShoppingCartOpen: boolean;
}

const initialState: State = {
  isSidebarOpen: false,
  isShoppingCartOpen: false,
}

export const uiSlice = createSlice({
  name: 'ui',
  initialState,
  reducers: {
    openSidebar: (state) => {
      state.isSidebarOpen = true
    },
    closeSidebar: (state) => {
      state.isSidebarOpen = false
    },
    closeShoppingCart: (state) => {
      state.isShoppingCartOpen = false
    },
    openShoppingCart: (state) => {
      state.isShoppingCartOpen = true
    }
  }
})

export const { closeSidebar, openSidebar, closeShoppingCart, openShoppingCart } = uiSlice.actions

export default uiSlice.reducer
