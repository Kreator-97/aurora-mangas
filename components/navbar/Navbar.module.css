
.container {
  background-image: url("/images/background.png");
  background-position: top;
  background-repeat: no-repeat;
  background-size: cover;
  background-attachment: fixed;
  height: var(--header-height);
  border-bottom: 1px solid var(--stroke-color);
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 100;
}

.content {
  display: grid;
  grid-template-columns: 1fr 32px 32px;
  column-gap: 1rem;
  row-gap: .5rem;
  grid-template-rows: 1fr 1fr;
  justify-content: space-between;
  align-items: center;
  padding: .5rem;
  max-width: var(--xl);
  margin: 0 auto;
}