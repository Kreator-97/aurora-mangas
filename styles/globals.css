@tailwind base;
@tailwind components;
@tailwind utilities;
@tailwind variants;

:root {
  box-sizing: border-box;
  --bg-color: #001E31;
  --bg-color2: #00474F;
  --stroke-color: #07695b70;
  --success: #21C728;
  --success-dark: #118A16;
  --dark: #333;
  --light: #eee;
  --header-height: 88px;
  --sm: 480px;
  --md: 720px;
  --lg: 960px;
  --xl: 1200px;
  --max: 1440px;
}

html {
  color: var(--light);
  background-image: linear-gradient( var(--bg-color), var(--bg-color2));
  background-position: center;
  background-repeat: no-repeat;
  background-size: cover;
  background-attachment: fixed;
}

.main {
  min-height: calc(100vh -var(--header-height));
  margin: var(--header-height) auto 0;
}

.title {
  @apply text-xl md:text-2xl lg:text-3xl text-center py-2;
}

.btn {
  @apply px-2 py-1 font-normal text-light rounded transition-colors border border-solid border-light;
}

.btn.ghost {
  @apply text-accent bg-opacity-0 border-accent;
}

.bg-success-gradient {
  background-image: linear-gradient( var(--success), var(--success-dark))!important;
}

.btn.ghost:hover {
  @apply bg-light;
}

.active {
  @apply bg-success;
}

@media screen and (min-width: 720px) {
  .main {
    padding: 0 .5rem;
  }
}
