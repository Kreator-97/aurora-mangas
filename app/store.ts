import { configureStore } from '@reduxjs/toolkit'
import { uiSlice, authSlice, shoppingCartSlice } from './slices'

export const store = configureStore({
  reducer: {
    ui: uiSlice,
    auth: authSlice,
    cart: shoppingCartSlice
  },
})

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch
