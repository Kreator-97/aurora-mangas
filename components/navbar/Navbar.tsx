import Image from 'next/image'
import { useRouter } from 'next/router'
import { HiMenu, HiOutlineShoppingCart, HiOutlineSearchCircle } from 'react-icons/hi'
import { useAppDispatch } from '../../app/hooks'
import { open } from '../../app/slices/uiSlice'
import styles from './Navbar.module.css'

export const Navbar = () => {
  const router = useRouter()
  const dispatch = useAppDispatch()
  
  const openMenu = () => {
    dispatch( open() )
  }

  const navigateTo = (url:string) => {
    router.push(url)
  } 

  return (
    <nav className={ styles.container }>
      <div className={ styles.content }>
        <Image
          src="/images/logo.png"
          alt="Aurora Mangas Logo"
          width={150}
          height={30}
          layout="fixed"
          className='cursor-pointer'
          onClick={ () => navigateTo('/') }
        />
        <HiOutlineShoppingCart
          color='white'
          size={32}
          cursor="pointer"/>
        <HiMenu
          color='white'
          size={32}
          cursor="pointer"
          onClick={ () => openMenu() }
        />
        <input
          type="search"
          placeholder='Encuentra tu manga favorito'
          className='col-span-2 rounded border-cyan-400 border p-1 text-sm text-dark'
        />
        <HiOutlineSearchCircle
          color='white'
          size={32}
          cursor="pointer"
        />
      </div>
    </nav>
  )
}
