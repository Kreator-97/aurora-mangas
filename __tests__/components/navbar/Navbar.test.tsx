import { expect } from '@jest/globals'
import { useRouter } from 'next/router'
import { Provider } from 'react-redux'
import { fireEvent, render, screen } from '@testing-library/react'

import { Navbar } from '../../../components/navbar/Navbar'
import { openSidebar, openShoppingCart } from '../../../app/slices/uiSlice'
import { store } from '../../../app/store'

const router = { query: { q: 'hello'}, push: jest.fn() }

jest.mock('next/router', () => ({
  useRouter: () => (router)
}))

jest.mock('../../../app/slices/uiSlice', () => ({
  openSidebar: jest.fn().mockReturnValue({
    payload: undefined, type: 'cart/openSidebar'
  }),
  openShoppingCart: jest.fn().mockReturnValue({
    payload: undefined, type: 'cart/openShoppingCart'
  })
}))

describe('test on Nabvar component', () => {
  test('should to match with snapshot', () => {
    
    const { container } = render(
      <Provider store={store}>
        <Navbar />
      </Provider>
    )
    expect(container).toMatchSnapshot()
  })

  test('openMenu function should to be called', () => {
    render(
      <Provider store={store}>
        <Navbar />
      </Provider>
    )
    const iconMenu = screen.getByTestId('menu')
    
    fireEvent.click(iconMenu)

    expect(openSidebar).toHaveBeenCalled()
  })

  test('openShoppingCart function should to be called', () => {
    render(
      <Provider store={store}>
        <Navbar />
      </Provider>
    )
    const cartMenu = screen.getByTestId('cart-menu')
    
    fireEvent.click(cartMenu)

    expect(openShoppingCart).toHaveBeenCalled()
  })

  test('onSearch function should to be called with query string', () => {
    render(
      <Provider store={store}>
        <Navbar />
      </Provider>
    )
    const searchForm = screen.getByTestId('search-form')
    const inputSearch = screen.getByRole('searchbox')
    
    const value = 'one punch man'
    fireEvent.change(inputSearch, {target: {value, name: 'search'}})

    fireEvent.submit(searchForm)
    expect(useRouter().push).toHaveBeenCalledWith(`/search?q=${value}`)
  })
})

