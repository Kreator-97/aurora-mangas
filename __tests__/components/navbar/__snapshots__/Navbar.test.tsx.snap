// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test on Nabvar component should to match with snapshot 1`] = `
<div>
  <nav
    class="container"
  >
    <div
      class="content"
    >
      <svg
        color="white"
        cursor="pointer"
        data-testid="menu"
        fill="none"
        height="32"
        stroke="currentColor"
        stroke-width="0.5"
        style="color: white;"
        viewBox="0 0 24 24"
        width="32"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4 6h16M4 12h16M4 18h16"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
      <span
        style="box-sizing: border-box; display: inline-block; overflow: hidden; width: 150px; height: 30px; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
      >
        <img
          alt="Aurora Mangas Logo"
          class="cursor-pointer"
          data-nimg="fixed"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </span>
      <div
        class="relative cursor-pointer"
        data-testid="cart-menu"
      >
        <svg
          color="white"
          cursor="pointer"
          fill="none"
          height="32"
          stroke="currentColor"
          stroke-width="0"
          style="color: white;"
          viewBox="0 0 24 24"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 3h2l.4 2M7 13h10l4-8H5.4M7 13L5.4 5M7 13l-2.293 2.293c-.63.63-.184 1.707.707 1.707H17m0 0a2 2 0 100 4 2 2 0 000-4zm-8 2a2 2 0 11-4 0 2 2 0 014 0z"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
        <span
          class="w-6 h-6 text-center font-semibold bg-accentDark bg-opacity-95 rounded-full absolute -top-2 -right-2"
        >
          0
        </span>
      </div>
      <form
        class="flex col-span-2"
        data-testid="search-form"
      >
        <input
          class="w-full rounded border-cyan-400 border p-1 text-sm text-dark"
          placeholder="Encuentra tu manga favorito. Ejemplo: one punch man #1"
          type="search"
          value="hello"
        />
      </form>
      <svg
        color="white"
        cursor="pointer"
        fill="none"
        height="32"
        stroke="currentColor"
        stroke-width="0"
        style="color: white;"
        viewBox="0 0 24 24"
        width="32"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 16l2.879-2.879m0 0a3 3 0 104.243-4.242 3 3 0 00-4.243 4.242zM21 12a9 9 0 11-18 0 9 9 0 0118 0z"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
  </nav>
</div>
`;
